clients/src/main/java/org/apache/kafka/common/requests/ProduceResponse.java

```diff
@@ -266,15 +266,14 @@ protected Struct toStruct(short version) {
                         .set(PARTITION_ID, partitionEntry.getKey())
                         .set(ERROR_CODE, errorCode)
                         .set(BASE_OFFSET_KEY_NAME, part.baseOffset);
-                if (partStruct.hasField(LOG_APPEND_TIME_KEY_NAME))
-                    partStruct.set(LOG_APPEND_TIME_KEY_NAME, part.logAppendTime);
+                partStruct.setIfExists(LOG_APPEND_TIME_KEY_NAME, part.logAppendTime);
                 partStruct.setIfExists(LOG_START_OFFSET_FIELD, part.logStartOffset);
 
                 List<Struct> errorRecords = new ArrayList<>();
-                for (Map.Entry<Integer, String> recordOffsetAndMessage : part.errorRecords.entrySet()) {
+                for (ErrorRecord recordOffsetAndMessage : part.errorRecords) {
                     Struct recordOffsetAndMessageStruct = partStruct.instance(ERROR_RECORDS_KEY_NAME)
-                            .set(RELATIVE_OFFSET_KEY_NAME, recordOffsetAndMessage.getKey())
-                            .setIfExists(RELATIVE_OFFSET_ERROR_MESSAGE_FIELD, recordOffsetAndMessage.getValue());
+                            .set(RELATIVE_OFFSET_KEY_NAME, recordOffsetAndMessage.getRelativeOffset())
+                            .setIfExists(RELATIVE_OFFSET_ERROR_MESSAGE_FIELD, recordOffsetAndMessage.getMessage());
```

Comment:
Not sure why we use `setIfExists` in this case, but not for relative offset.

Reply:
ack